sca: codechecker: Twister default values

pdgendt opened the pull request:
This PR enables running Static Code Analysis tools as part of twister. The main focus is to have a [CodeChecker](https://codechecker.readthedocs.io/en/latest/) server running to track reported issues.

An example:
```shell
# Setup the environment
$ export CODECHECKER_CONFIG_FILE="$PWD/.codechecker.yml"
$ export CODECHECKER_STORE_OPTS="--url;localhost:8001/Zephyr"
$ export CODECHECKER_CLEANUP=y
$ export CODECHECKER_PARSE_SKIP=y
# Run some twisters
$ west twister -T samples/net/sockets -p native_sim --build-only -x=ZEPHYR_SCA_VARIANT=codechecker
$ west twister -T samples/hello_world -p mimxrt1064_evk -p qemu_x86 --build-only -x=ZEPHYR_SCA_VARIANT=codechecker
```

![image](https://github.com/user-attachments/assets/59a3d095-e7e2-4762-ab19-f894ff18d3a8)
![image](https://github.com/user-attachments/assets/5e91a59d-74a4-4d2c-beaf-b1a9e9d120c9)
![image](https://github.com/user-attachments/assets/3039cbfe-7a90-4c79-a0b2-869b7407c8da)

pdgendt committed:
cmake: sca: codechecker: Use zephyr_get for variables

Allow setting CodeChecker cmake variables from multiple sources with
zephyr_get.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
cmake: sca: codechecker: Configurable variables

Add a separate variables for CodeChecker commands

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
doc: develop: sca: Update CodeChecker configuration options

Put CodeChecker configuration parameters in tables for a clear overview
of all the options.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
tests: Add a CodeChecker config file

Create a file to be used in CI or locally that lists all enabled/disabled
checks.

Can be passed as a CMake argument or set as environment variable:
CODECHECKER_CONFIG_FILE=$ZEPHYR_BASE/tests/codechecker_config.yaml

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt commented on doc/develop/sca/codechecker.rst:
```diff
@@ -33,56 +33,62 @@ called with a ``-DZEPHYR_SCA_VARIANT=codechecker`` parameter, e.g.
+
+When running ``twister``, it is possible to run CodeChecker static code analysis as well. Pass
+``--sca-variant codechecker`` when calling ``twister`` to enable it. Once enabled, some default
+options are set as following:
```
Tried this, but referring from the docs:

```
Exit status
------------------------------------------------
0 - No report
1 - CodeChecker error
2 - At least one report emitted by an analyzer
```

There will almost always be some report, so not sure how useful this is in twister.
The same can be said for code coverage, should you "fail" if a coverage threshold isn't reached?

golowanow replied:
My point is to have the same behavior for all SCA in Twister with default options, but currently this PR results differently:
* with `--sca-variant=gcc` there are build warnings from SCA, eventually Twister test instance build error.
* with `--sca-variant=codechecker` issues become errors only with additional `CODECHECKER_PARSE_EXIT_STATUS=y`

This should be, at least, documented in the `--sca-variant` description.

tejlmand commented on cmake/sca/codechecker/sca.cmake:
```diff
@@ -5,6 +5,40 @@
 message(STATUS "Found SCA: CodeChecker (${CODECHECKER_EXE})")
 
+# Get CodeChecker specific variables
+zephyr_get(CODECHECKER_ANALYZE_JOBS)
```
I know each SCA tool will have its own prefered way of being configured, but at the same time Zephyr users are familiar with Kconfig.

During review of Eclair integration it became clear that a better Kconfig integration to SCA tools is needed, see: #68324 and https://github.com/zephyrproject-rtos/zephyr/issues/77018, especially in a way that the SCA Kconfig is only loaded when the tool is available.

The list of settings below looks like something that would be better placed in Kconfig, and thereby also be auto documented using the Kconfig help text.

Perhaps with an exception for the `CODECHECKER_CONFIG_FILE` as that could look like something one might want to pass in directly on CMake invocation.

Thoughts ?

tejlmand commented on scripts/pylib/twister/twisterlib/sca.py:
```diff
@@ -0,0 +1,34 @@
+# vim: set syntax=python ts=4 :
```
I don't think we generally add editor settings in files, although I give a :+1: for the choice of editor.

```suggestion
```

kartben replied:
+1 - I think the accepted practice is just to run files through ``black file.py -l 100``

tejlmand reviewed (commented):
Thanks for this.

Some comments, but so far no hard blockers.

pdgendt committed:
cmake: sca: codechecker: CMake targets instead of commands

Replace CodeChecker post build commands with regular CMake Targets.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
scripts: pylib: twister: Pass testsuite name to cmake

Add a CMake argument with the current testsuite name.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
cmake: sca: codechecker: Parse twister variables

If we're running in a twister context, use the information as sane
defaults for CodeChecker.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
cmake: Create git module

Create a CMake git module for a git_describe function.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
cmake: sca: codechecker: Add cleanup step

Add the option to remove the static code analysis reports after
parsing/storing.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

nashif commented on tests/codechecker_config.yaml:
```diff
@@ -0,0 +1,21 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# Copyright (c) 2024, Basalte bv
```
i would put this file somewhere else, the config might be applied when building tests, but we are checking the zephyr code, not the tests only, maybe a top level dot file or directory with multiple configs?
